### PR TITLE
Dilpreet | Stage | Pending Related Voucher

### PR DESCRIPTION
**Pull request by dilpreetsinghofficial**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Other information**:

### CODE REVIEW

File: apps/web-giddh/src/app/vouchers/create/create.component.ts [hunk 1]
```diff
@@ -3488,7 +3488,7 @@ export class VoucherCreateComponent implements OnInit, OnDestroy, AfterViewInit
         this.invoiceForm.get('updateAccountDetails')?.patchValue(false);
         if (this.isPendingEntries) {
             this.saveVoucher(() => {
-                this.router.navigate(['/pages/vouchers/preview/sales/pending']);
+                this.router.navigate([`/pages/vouchers/preview/${this.queryParams.voucherType}/pending`]);
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/vouchers/create/create.component.ts:3491]: check this for all voucher

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 2]
```diff
@@ -297,6 +299,35 @@ export class VoucherListComponent implements OnInit, OnDestroy {
     };
     /** Holds current route query parameters */
     public queryParams: any = {};
+    /** True if voucher generate in process */
+    public generateVoucherInProcess: boolean = false;
+    /** Decimal places from company settings */
+    public giddhBalanceDecimalPlaces: number = 2;
+    /** Duplicate copy of entry unique names for bulk action variable */
+    public entryUniqueNamesForBulkActionDuplicateCopy: GenerateBulkInvoiceObject[] = [];
+    /** selected pending voucher */
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]:
```suggestion
    /** Selected pending voucher */
```

Do this for all

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 3]
```diff
@@ -297,6 +299,35 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+    public customDateSelected: boolean = false;
+    /** Instance of ledger search request */
+    public ledgerSearchRequest: InvoiceFilterClass = new InvoiceFilterClass();
+    /** Loading Observable */
+    public isGetAllRequestInProcess$: Observable<any> = this.componentStore.getLedgerDataInProcess$;
+    /** Holds Ledger Data */
+    public ledgersData: any[] = [];
+    /** Holds voucher type for credit/debit note*/
+    public voucherTypes: any = [
+        { label: 'Credit Note', value: 'credit note' },
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]: Use translated label

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 4]
```diff
@@ -608,6 +651,64 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+                }
+                if (this.activeModule === 'pending') {
+                    this.getLedgersOfInvoice();
+                }
+            }
+        });
+
+        this.componentStore.pendingVoucherList$.pipe(takeUntil(this.destroyed$)).subscribe((res: GetAllLedgersForInvoiceResponse) => {
+            if (res && res.results) {
+
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]:
```suggestion
```

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 5]
```diff
@@ -771,10 +872,14 @@ export class VoucherListComponent implements OnInit, OnDestroy {
         } else if (this.activeTabGroup === 3) {
             if (this.voucherType === 'receipt' && this.activeModule === 'list') {
                 this.selectedTabIndex = 0;
+            } else if (this.voucherType === 'receipt' && this.activeModule === 'pending') {
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]: use enum if possible

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 6]
```diff
@@ -771,10 +872,14 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+                this.selectedTabIndex = 1;
             }
         } else if (this.activeTabGroup === 4) {
             if (this.voucherType === 'payment' && this.activeModule === 'list') {
                 this.selectedTabIndex = 0;
+            } else if (this.voucherType === 'payment' && this.activeModule === 'pending') {
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]: use enum if possible

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 7]
```diff
@@ -771,10 +872,14 @@ export class VoucherListComponent implements OnInit, OnDestroy {
         } else if (this.activeTabGroup === 3) {
             if (this.voucherType === 'receipt' && this.activeModule === 'list') {
                 this.selectedTabIndex = 0;
+            } else if (this.voucherType === 'receipt' && this.activeModule === 'pending') {
+                this.selectedTabIndex = 1;
             }
         } else if (this.activeTabGroup === 4) {
             if (this.voucherType === 'payment' && this.activeModule === 'list') {
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]: use enum if possible

Here to

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 8]
```diff
@@ -1938,4 +2121,260 @@ export class VoucherListComponent implements OnInit, OnDestroy {
             this.router.navigate([`/pages/vouchers/${this.urlVoucherType}/${voucher?.account?.uniqueName ?? voucher?.vendor?.uniqueName}/${voucher?.uniqueName}/copy`]);
         }
     }
-}
\ No newline at end of file
+    /**
+     *Fetch ledgers for invoices.
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]:
```suggestion
     * Fetch ledgers for invoices.
```

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 9]
```diff
@@ -1938,4 +2121,260 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+     *
+     * @memberof VoucherListComponent
+     */
+    public previewInvoice(): void {
+        const voucher = this.selectedPendingVouchers[0];
+        const uniqueNames = this.selectedPendingVouchers.map(voucher => voucher.uniqueName).join(',');
+        if (voucher) {
+            this.router.navigate(
+                [`/pages/vouchers/${voucher.voucherType}/${voucher.account?.uniqueName}/create`],
+                { queryParams: { entryUniqueNames: uniqueNames, voucherType:this.voucherType } }
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]:
```suggestion
                { queryParams: { entryUniqueNames: uniqueNames, voucherType: this.voucherType } }
```

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 10]
```diff
@@ -1938,4 +2121,260 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+     *  Navigates to the preview invoice page.
+     *
+     * @memberof VoucherListComponent
+     */
+    public previewInvoice(): void {
+        const voucher = this.selectedPendingVouchers[0];
+        const uniqueNames = this.selectedPendingVouchers.map(voucher => voucher.uniqueName).join(',');
+        if (voucher) {
+            this.router.navigate(
+                [`/pages/vouchers/${voucher.voucherType}/${voucher.account?.uniqueName}/create`],
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts:2243]:
```suggestion
                [`/pages/vouchers/${voucher.voucherType ?? this.voucherType}/${voucher.account?.uniqueName}/create`],
```

File: apps/web-giddh/src/app/vouchers/list/list.component.html [hunk 11]
```diff
@@ -1173,114 +1702,236 @@ <h3>{{ element?.customerName | titlecase }}</h3>
...
+                                    </td>
+                                </ng-container>
+                                <ng-container matColumnDef="account">
+                                    <th mat-header-cell *matHeaderCellDef>{{ commonLocaleData?.app_account }}</th>
+                                    <td mat-cell *matCellDef="let element" class="no-dbl-click-select">
+                                        <div
+                                            class="cursor-pointer"
+                                            (dblclick)="!isCompany && (element.showAccountDetails = true)"
+                                        >
+                                            {{ element?.account?.name | titlecase }}
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/vouchers/list/list.component.html:1849]:
```suggestion
                                            {{ element.account?.name | titlecase }}
```

Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vouchers/list/list.component.html:1849]: check every where

**Review by dvCodeWorld — CHANGES_REQUESTED** [review]

**Review by dvCodeWorld — DISMISSED** [review]

File: apps/web-giddh/src/app/vouchers/create/create.component.ts (hunk 1 above)
Comment (dilpreetsinghofficial) [reply apps/web-giddh/src/app/vouchers/create/create.component.ts:3491]: this is working for only sales, credit note , debit note etc

File: apps/web-giddh/src/app/vouchers/list/list.component.ts (hunk 10 above)
Comment (dilpreetsinghofficial) [reply apps/web-giddh/src/app/vouchers/list/list.component.ts:2243]: No need

**Review by dvCodeWorld — DISMISSED** [review]

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 12]
```diff
@@ -1938,4 +2117,276 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+        this.entryUniqueNamesForBulkActionDuplicateCopy = cloneDeep(model);
+        this.store.dispatch(
+            this.invoiceActions.GenerateBulkInvoice(
+                { combined: action },
+                { entryUniqueNames: model, generateEInvoice }
+            )
+        );
+
+        this.selectedPendingVouchers = [];
+        return true;
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]: remove useless return statement

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 13]
```diff
@@ -1938,4 +2117,276 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+    }
+
+    /**
+     * Groups pending vouchers by account unique name.
+     *
+     * @private
+     * @return {*}  {Record<string, GenBulkInvoiceGroupByObj[]>}
+     * @memberof VoucherListComponent
+     */
+    private groupPendingVouchersByAccount(): Record<string, GenBulkInvoiceGroupByObj[]> {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts:2285]: inform usage of Record

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 14]
```diff
@@ -1938,4 +2117,276 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+
+    /**
+     *  Flattens grouped vouchers into an array of unique names.
+     *
+     * @private
+     * @param {Record<string, GenBulkInvoiceGroupByObj[]>} grouped
+     * @return {*}  {string[]}
+     * @memberof VoucherListComponent
+     */
+    private flattenGroupedVouchers(grouped: Record<string, GenBulkInvoiceGroupByObj[]>): string[] {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]: rename parameters

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 15]
```diff
@@ -1938,4 +2117,276 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+    /**
+     * Callback for translation response complete
+     *
+     * @param {boolean} event
+     * @memberof RatioAnalysisChartComponent
+     */
+    public translationComplete(event: boolean): void {
+        if (event) {
+            this.voucherTypes = [
+                { label: this.localeData.tabs.credit_note, value: 'credit note' },
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]: use voucher type enum

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 16]
```diff
@@ -1938,4 +2117,276 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+            voucher?.account?.uniqueName === firstAccountUniqueName
+        );
+    }
+
+    /**
+     *  Navigates to the preview invoice page.
+     *
+     * @memberof VoucherListComponent
+     */
+    public previewInvoice(): void {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]: correct method name based on usage

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 17]
```diff
@@ -1938,4 +2117,276 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+        this.isUniversalDateApplicable = false;
+    }
+
+    /**
+     * This will be check account of same accouunt on selected pending vouchers
+     *
+     * @return {*}  {boolean}
+     * @memberof VoucherListComponent
+     */
+    public isSameAccount(): boolean {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]: rename method

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 18]
```diff
@@ -1938,4 +2117,276 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+        return !this.ledgersData || this.ledgersData.length === 0;
+    }
+
+    /**
+     * Decrements the page if it's greater than 1.
+     *
+     * @private
+     * @memberof VoucherListComponent
+     */
+    private decrementPageIfNeeded(): void {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts:2151]: mention the usage of this logic

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 19]
```diff
@@ -1938,4 +2117,276 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+        this.selectedPendingVouchers = [];
+    }
+
+    /**
+     * Dispatches a request to fetch ledgers.
+     *
+     * @private
+     * @memberof VoucherListComponent
+     */
+    private dispatchLedgerRequest(): void {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]: rename method names

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 20]
```diff
@@ -608,6 +648,63 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+        });
+
+        // listen for bulk invoice generate and successfully generate and do the things
+        this.componentStore.isBulkInvoiceGenerated$.subscribe(result => {
+            if (result) {
+                this.selectAllPendingVouchers(false);
+                this.getLedgersOfInvoice();
+            }
+        });
+        this.componentStore.isBulkInvoiceGeneratedWithoutError$.subscribe(result => {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]: explain usage

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 21]
```diff
@@ -608,6 +648,63 @@ export class VoucherListComponent implements OnInit, OnDestroy {
...
+                    });
+                }
+                this.ledgersData = response?.results;
+            }
+        });
+
+        // listen for bulk invoice generate and successfully generate and do the things
+        this.componentStore.isBulkInvoiceGenerated$.subscribe(result => {
+            if (result) {
+                this.selectAllPendingVouchers(false);
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]:
```suggestion
                this.selectAllPendingVouchers({ checked: false });
```

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 22]
```diff
@@ -312,7 +340,9 @@ export class VoucherListComponent implements OnInit, OnDestroy {
         private invoiceService: InvoiceService,
         private adjustmentUtilityService: AdjustmentUtilityService,
         private salesAction: SalesActions,
-        private settingsProfileActions: SettingsProfileActions
+        private settingsProfileActions: SettingsProfileActions,
+        private invoiceActions: InvoiceActions,
+        private commonActions: CommonActions
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts]: remove useless import

File: apps/web-giddh/src/app/vouchers/list/list.component.ts [hunk 23]
```diff
@@ -57,10 +59,6 @@ export interface PeriodicElementPending {
     total: string;
     description: string;
 }
-// pending-table
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.ts:60]: remove unused interface also

File: apps/web-giddh/src/app/vouchers/list/list.component.html [hunk 24]
```diff
@@ -1173,114 +1703,236 @@ <h3>{{ element?.customerName | titlecase }}</h3>
...
+                                    <td mat-cell *matCellDef="let element">
+                                        {{ element?.entryDate }}
+                                    </td>
+                                </ng-container>
+                                <ng-container matColumnDef="particular">
+                                    <th mat-header-cell *matHeaderCellDef>{{ commonLocaleData?.app_particular }}</th>
+                                    <td mat-cell *matCellDef="let element" class="p-0">
+                                        <ng-container *ngFor="let transaction of element?.transactions">
+                                            <p>{{ transaction?.particular?.name }}</p>
+                                            <p *ngIf="transaction?.inventory && transaction.inventory?.stock">
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.html]:
```suggestion
                                            <p *ngIf="transaction?.inventory?.stock">
```

File: apps/web-giddh/src/app/vouchers/list/list.component.html [hunk 25]
```diff
@@ -1173,114 +1703,236 @@ <h3>{{ element?.customerName | titlecase }}</h3>
...
+                                    <th mat-header-cell *matHeaderCellDef class="text-right">
+                                        {{ commonLocaleData?.app_amount }}
+                                    </th>
+                                    <td mat-cell *matCellDef="let element" class="text-right p-0">
+                                        <ng-container *ngFor="let transaction of element?.transactions">
+                                            <p>
+                                                <amount-field
+                                                    [amount]="transaction?.amount"
+                                                    [currencySymbol]="element.accountCurrencySymbol"
+                                                    [currencyCode]="activeCompany?.baseCurrency"
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.html]: use account base currency

File: apps/web-giddh/src/app/vouchers/list/list.component.html [hunk 26]
```diff
@@ -1173,114 +1703,236 @@ <h3>{{ element?.customerName | titlecase }}</h3>
...
+                                            class="outline-card account-detail-wrapper"
+                                            *ngIf="element?.showAccountDetails"
+                                            (clickOutside)="element.showAccountDetails = false"
+                                        >
+                                            <div
+                                                class="d-flex align-items-center justify-content-between column-gap15 p-3 border-bottom account-header"
+                                            >
+                                                <h3>{{ element?.account?.name }}</h3>
+                                                <span class="cursor-pointer" (click)="editAccount(element)">
+                                                    <img src="./assets/images/path.svg" />
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.html]: use correct image path to handle electron

File: apps/web-giddh/src/app/vouchers/list/list.component.html [hunk 27]
```diff
@@ -1173,114 +1703,236 @@ <h3>{{ element?.customerName | titlecase }}</h3>
...
+                                        *matCellDef="let element"
+                                        class="text-right"
+                                        matTooltipPosition="above"
+                                        [matTooltip]="element?.totalTooltipText"
+                                    >
+                                        <span class="d-inline-flex">
+                                            <amount-field
+                                                [amount]="element.total.amount"
+                                                [currencySymbol]="element.accountCurrencySymbol"
+                                                [currencyCode]="activeCompany?.baseCurrency"
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vouchers/list/list.component.html]: use account currency

**Review by ravinderwalkover — CHANGES_REQUESTED** [review]

File: apps/web-giddh/src/app/vouchers/list/list.component.ts (hunk 13 above)
Comment (dilpreetsinghofficial) [reply apps/web-giddh/src/app/vouchers/list/list.component.ts:2285]: Record is the use for key and values where we know what the type of key and value is.

File: apps/web-giddh/src/app/vouchers/list/list.component.ts (hunk 18 above)
Comment (dilpreetsinghofficial) [reply apps/web-giddh/src/app/vouchers/list/list.component.ts:2151]: This is the use case for when decrement page automatically when records are deleted or generated

File: apps/web-giddh/src/app/vouchers/list/list.component.ts (hunk 20 above)
Comment (dilpreetsinghofficial) [reply apps/web-giddh/src/app/vouchers/list/list.component.ts]: removed

**Review by ravinderwalkover — APPROVED** [review]

**Pull request merged**